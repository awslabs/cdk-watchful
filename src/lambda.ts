import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct, Duration } from '@aws-cdk/core';
import { IWatchful } from './api';

const DEFAULT_DURATION_THRESHOLD_PERCENT = 80;

export interface WatchLambdaFunctionOptions {
  /**
   * Flag to disable alerting on errors
   *
   * @default false
   */
  readonly errorsDisableAlerts?: boolean;

  /**
   * Number of allowed errors per minute. If there are more errors than that, an alarm will trigger.
   *
   * @default 0
   */
  readonly errorsPerMinuteThreshold?: number;

  /**
   * Flag to enable alerting on invocationsMetric
   *
    @default false
   */
  readonly invocationsEnableAlerts?: boolean;

  /**
   * Threshold for alerting invocations
   */
  readonly invocationsThreshold?: Duration;

  /**
   * Number of allowed throttles per minute.
   *
   * @default 0
   */
  readonly throttlesPerMinuteThreshold?: number;

  /**
   * Threshold for the duration alarm as percentage of the function's timeout
   * value.
   *
   * If this is set to 50%, the alarm will be set when p99 latency of the
   * function exceeds 50% of the function's timeout setting.
   *
   * @default 80
   */
  readonly durationThresholdPercent?: number;

  /**
   * Override duration timeout threshold.
   * Necessary for lambdas that aren't created via the CDK.
   * This value is still adjusted by durationThresholdPercent
   *
   * @default 3
   */
  readonly durationTimeoutSec?: number;
}

export interface WatchLambdaFunctionProps extends WatchLambdaFunctionOptions {
  readonly title: string;
  readonly watchful: IWatchful;
  readonly fn: lambda.IFunction;
}

export class WatchLambdaFunction extends Construct {

  private readonly watchful: IWatchful;
  private readonly fn: lambda.IFunction;
  private invocationsMetric!: cloudwatch.Metric;
  private invocationsAlarm: cloudwatch.Alarm | undefined;
  private errorsMetric!: cloudwatch.Metric;
  private errorsAlarm: cloudwatch.Alarm | undefined;
  private throttlesMetric!: cloudwatch.Metric;
  private throttlesAlarm!: cloudwatch.Alarm;
  private durationMetric!: cloudwatch.Metric;
  private durationAlarm!: cloudwatch.Alarm;

  constructor(scope: Construct, id: string, props: WatchLambdaFunctionProps) {
    super(scope, id);

    const cfnFunction = props.fn.node.defaultChild as lambda.CfnFunction;
    // Use the parameter if it was passed, then the lambdas timeout if it exists, lastly use the default of 3 seconds
    const timeoutSec = props.durationTimeoutSec || cfnFunction?.timeout || 3;

    this.watchful = props.watchful;
    this.fn = props.fn;

    this.watchful.addSection(props.title, {
      links: [
        { title: 'AWS Lambda Console', url: linkForLambdaFunction(this.fn) },
        { title: 'CloudWatch Logs', url: linkForLambdaLogs(this.fn) },
      ],
    });

    this.createInvocationsMonitor(props.invocationsEnableAlerts, props.invocationsThreshold);
    this.createErrorsMonitor(props.errorsDisableAlerts, props.errorsPerMinuteThreshold);
    this.createThrottlesMonitor(props.throttlesPerMinuteThreshold);
    this.createDurationMonitor(timeoutSec!, props.durationThresholdPercent);

    let invocationWidget: cloudwatch.IWidget;
    if (!props.invocationsEnableAlerts) {
      invocationWidget = new cloudwatch.GraphWidget({
        title: `Invocations/${this.invocationsMetric.period.toMinutes()}min`,
        width: 6,
        left: [this.invocationsMetric],
      });
    } else {
      invocationWidget = new cloudwatch.AlarmWidget({
        title: `Invocations/${this.invocationsMetric.period.toMinutes()}min`,
        alarm: this.invocationsAlarm!,
        width: 6,
      });
    }

    let errorWidget: cloudwatch.IWidget;
    if (props.errorsDisableAlerts) {
      errorWidget = new cloudwatch.GraphWidget({
        title: `Errors/${this.errorsMetric.period.toMinutes()}min`,
        width: 6,
        left: [this.errorsMetric],
      });
    } else {
      errorWidget = new cloudwatch.AlarmWidget({
        title: `Errors/${this.errorsMetric.period.toMinutes()}min`,
        alarm: this.errorsAlarm!,
        width: 6,
      });
    }

    this.watchful.addWidgets(
      invocationWidget,
      errorWidget,
      new cloudwatch.GraphWidget({
        title: `Throttles/${this.throttlesMetric.period.toMinutes()}min`,
        width: 6,
        left: [this.throttlesMetric],
        leftAnnotations: [this.throttlesAlarm.toAnnotation()],
      }),
      new cloudwatch.GraphWidget({
        title: `Duration/${this.durationMetric.period.toMinutes()}min`,
        width: 6,
        left: [this.durationMetric],
        leftAnnotations: [this.durationAlarm.toAnnotation()],
      }),
    );
  }

  createInvocationsMonitor(invocationsEnableAlerts = false, invocationsThreshold = Duration.hours(1)) {
    const fn = this.fn;
    this.invocationsMetric = fn.metricInvocations();
    if (invocationsEnableAlerts) {
      this.invocationsAlarm = this.invocationsMetric.createAlarm(this, 'InvocationAlarm', {
        alarmDescription: `Expecting invocations to occur every ${invocationsThreshold.toHours()}hours`,
        threshold: 1,
        period: invocationsThreshold, // deprecated but functional
        comparisonOperator: cloudwatch.ComparisonOperator.LESS_THAN_THRESHOLD,
        evaluationPeriods: 1,
        treatMissingData: cloudwatch.TreatMissingData.BREACHING,
      });
      // this isn't working as promised, so keep using the deprecated method for now
      // this.invocationsMetric.with({ period: invocationsThreshold } );
      this.watchful.addAlarm(this.invocationsAlarm);
    }
  }

  private createErrorsMonitor(errorsDisableAlerts = false, errorsPerMinuteThreshold = 0) {
    const fn = this.fn;
    this.errorsMetric = fn.metricErrors();
    if (!errorsDisableAlerts) {
      this.errorsAlarm = this.errorsMetric.createAlarm(this, 'ErrorsAlarm', {
        alarmDescription: `Over ${errorsPerMinuteThreshold} errors per minute`,
        threshold: errorsPerMinuteThreshold,
        comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
        evaluationPeriods: 3,
      });
      this.watchful.addAlarm(this.errorsAlarm);
    }
  }

  private createThrottlesMonitor(throttlesPerMinuteThreshold = 0) {
    const fn = this.fn;
    this.throttlesMetric = fn.metricThrottles();
    this.throttlesAlarm = this.throttlesMetric.createAlarm(this, 'ThrottlesAlarm', {
      alarmDescription: `Over ${throttlesPerMinuteThreshold} throttles per minute`,
      threshold: throttlesPerMinuteThreshold,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
      evaluationPeriods: 3,
    });
    this.watchful.addAlarm(this.throttlesAlarm);
  }

  private createDurationMonitor(timeoutSec: number, durationPercentThreshold: number = DEFAULT_DURATION_THRESHOLD_PERCENT) {
    const fn = this.fn;
    this.durationMetric = fn.metricDuration();
    const durationThresholdSec = Math.floor(durationPercentThreshold / 100 * timeoutSec);
    this.durationAlarm = this.durationMetric.createAlarm(this, 'DurationAlarm', {
      alarmDescription: `p99 latency >= ${durationThresholdSec}s (${durationPercentThreshold}%)`,
      comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
      threshold: durationThresholdSec * 1000, // milliseconds
      evaluationPeriods: 3,
    });
    this.watchful.addAlarm(this.durationAlarm);
  }
}

function linkForLambdaFunction(fn: lambda.IFunction, tab = 'graph') {
  return `https://console.aws.amazon.com/lambda/home?region=${fn.stack.region}#/functions/${fn.functionName}?tab=${tab}`;
}

function linkForLambdaLogs(fn: lambda.IFunction) {
  return `https://console.aws.amazon.com/cloudwatch/home?region=${fn.stack.region}#logEventViewer:group=/aws/lambda/${fn.functionName}`;
}