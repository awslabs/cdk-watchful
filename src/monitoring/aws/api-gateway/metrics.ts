import { Metric, Statistic } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';
import { WatchedOperation } from '../../../api-gateway';

const enum Metrics {
  FourHundredError = '4XXError',
  FiveHundredError = '5XXError',
  CacheHitCount = 'CacheHitCount',
  CacheMissCount = 'CacheMissCount',
  Count = 'Count',
  IntegrationLatency = 'IntegrationLatency',
  Latency = 'Latency',
}

const Namespace = 'AWS/ApiGateway';

export class ApiGatewayMetricFactory {
  metricErrors(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      errors4XX: this.metric(Metrics.FourHundredError, apiName, stage, op).with({ label: '4XX Errors', statistic: Statistic.SUM }),
      errors5XX: this.metric(Metrics.FiveHundredError, apiName, stage, op).with({ label: '5XX Errors', statistic: Statistic.SUM }),
    };
  }

  metricCache(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      hits: this.metric(Metrics.CacheHitCount, apiName, stage, op).with({ label: 'Cache Hit', statistic: Statistic.SUM }),
      misses: this.metric(Metrics.CacheMissCount, apiName, stage, op).with({ label: 'Cache Miss', statistic: Statistic.SUM }),
    };
  }

  metricCalls(apiName: string, stage: string, op?: WatchedOperation) {
    return this.metric(Metrics.Count, apiName, stage, op).with({ label: 'Calls', statistic: Statistic.SUM });
  }

  metricIntegrationLatency(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      min: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'min', statistic: Statistic.MINIMUM }),
      avg: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'avg', statistic: Statistic.AVERAGE }),
      p90: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p90', statistic: 'p90' }),
      p95: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p95', statistic: 'p95' }),
      p99: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'p99', statistic: 'p99' }),
      max: this.metric(Metrics.IntegrationLatency, apiName, stage, op).with({ label: 'max', statistic: Statistic.MAXIMUM }),
    };
  }

  metricLatency(apiName: string, stage: string, op?: WatchedOperation) {
    return {
      min: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'min', statistic: Statistic.MINIMUM }),
      avg: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'avg', statistic: Statistic.AVERAGE }),
      p90: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p90', statistic: 'p90' }),
      p95: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p95', statistic: 'p95' }),
      p99: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'p99', statistic: 'p99' }),
      max: this.metric(Metrics.Latency, apiName, stage, op).with({ label: 'max', statistic: Statistic.MAXIMUM }),
    };
  }

  protected metric(metricName: Metrics, apiName: string, stage: string, op?: WatchedOperation) {
    return new Metric({
      metricName,
      namespace: Namespace,
      period: Duration.minutes(1),
      dimensions: {
        ApiName: apiName,
        Stage: stage,
        ...op && {
          Method: op.httpMethod,
          Resource: op.resourcePath,
        },
      },
    });
  }
}

