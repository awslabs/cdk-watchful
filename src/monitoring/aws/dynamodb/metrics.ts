import { Metric } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';

export class DynamoDbMetricFactory {
  static readonly Namespace = 'AWS/DynamoDB';

  metricConsumedCapacityUnits(tableName: string) {
    function metric(metricName: string, label: string) {
      return new Metric({
        metricName,
        label,
        namespace: DynamoDbMetricFactory.Namespace,
        period: Duration.minutes(1),
        statistic: 'Sum',
        dimensions: {
          TableName: tableName,
        },
      });
    }

    return {
      read: metric('ConsumedReadCapacityUnits', 'Consumed (Read)'),
      write: metric('ConsumedWriteCapacityUnits', 'Consumed (Write)'),
    };
  }
}
