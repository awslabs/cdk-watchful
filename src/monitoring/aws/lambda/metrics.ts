import { Metric } from '@aws-cdk/aws-cloudwatch';
import * as cdk from '@aws-cdk/core';

export class LambdaMetricFactory {
  static readonly Namespace = 'AWS/Lambda';

  metricInvocations(functionName: string) {
    return new Metric({
      metricName: 'Invocations',
      namespace: LambdaMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        FunctionName: functionName,
      },
    });
  }

  metricDuration(functionName: string) {
    function metric(statistic: string) {
      return new Metric({
        metricName: 'Duration',
        label: statistic,
        namespace: LambdaMetricFactory.Namespace,
        period: cdk.Duration.minutes(5),
        statistic,
        dimensions: {
          FunctionName: functionName,
        },
      });
    }

    return {
      min: metric('Minimum'),
      avg: metric('Average'),
      p50: metric('p50'),
      p90: metric('p90'),
      p99: metric('p99'),
      max: metric('Maximum'),
    };
  }

  metricErrors(functionName: string) {
    return new Metric({
      metricName: 'Errors',
      namespace: LambdaMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        FunctionName: functionName,
      },
    });
  }

  metricThrottles(functionName: string) {
    return new Metric({
      metricName: 'Throttles',
      namespace: LambdaMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        FunctionName: functionName,
      },
    });
  }
}
