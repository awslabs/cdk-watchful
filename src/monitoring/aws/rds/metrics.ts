import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import * as cdk from '@aws-cdk/core';

/**
 * Metrics for RDS Aurora.
 */
export class RdsAuroraMetricFactory {
  static readonly Namespace = 'AWS/RDS';

  metricDmlThroughput(clusterIdentifier: string) {
    const dbInsertThroughputMetric = new cloudwatch.Metric({
      metricName: 'InsertThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    const dbUpdateThroughputMetric = new cloudwatch.Metric({
      metricName: 'UpdateThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    const dbSelectThroughputMetric = new cloudwatch.Metric({
      metricName: 'SelectThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    const dbDeleteThroughputMetric = new cloudwatch.Metric({
      metricName: 'DeleteThroughput',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Sum',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
    return {
      dbInsertThroughputMetric,
      dbUpdateThroughputMetric,
      dbSelectThroughputMetric,
      dbDeleteThroughputMetric,
    };
  }

  metricBufferCacheHitRatio(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'BufferCacheHitRatio',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }

  metricDbConnections(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'DatabaseConnections',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }

  metricReplicaLag(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'AuroraReplicaLag',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }

  metricCpuUtilization(clusterIdentifier: string) {
    return new cloudwatch.Metric({
      metricName: 'CPUUtilization',
      namespace: RdsAuroraMetricFactory.Namespace,
      period: cdk.Duration.minutes(5),
      statistic: 'Average',
      dimensions: {
        DBClusterIdentifier: clusterIdentifier,
      },
    });
  }
}
