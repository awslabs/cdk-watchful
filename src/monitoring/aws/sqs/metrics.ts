import { Metric, Statistic } from '@aws-cdk/aws-cloudwatch';
import { Duration } from '@aws-cdk/core';

const enum Metrics {
  ApproximateNumberOfMessagesVisible = 'ApproximateNumberOfMessagesVisible',
  NumberOfMessagesSent = 'NumberOfMessagesSent',
  ApproximateAgeOfOldestMessage = 'ApproximateAgeOfOldestMessage',
  SentMessageSize = 'SentMessageSize'
}

const Namespace = 'AWS/SQS';

export class SqsMetricFactory {
  metricApproximateVisibleMessages(topicName: string) {
    return this.metric(Metrics.ApproximateNumberOfMessagesVisible, topicName).with({ statistic: Statistic.MAXIMUM });
  }

  metricIncomingMessages(topicName: string) {
    return this.metric(Metrics.NumberOfMessagesSent, topicName).with({ statistic: Statistic.SUM });
  }

  metricAgeOfOldestMessageInSeconds(topicName: string) {
    return this.metric(Metrics.ApproximateAgeOfOldestMessage, topicName).with({ statistic: Statistic.MAXIMUM });
  }

  metricAverageMessageSizeInBytes(topicName: string) {
    return this.metric(Metrics.SentMessageSize, topicName).with({ statistic: Statistic.AVERAGE });
  }

  protected metric(metric: Metrics, topicName: string) {
    return new Metric({
      metricName: metric,
      namespace: Namespace,
      period: Duration.minutes(5),
      dimensions: {
        TopicName: topicName,
      },
    });
  }
}
