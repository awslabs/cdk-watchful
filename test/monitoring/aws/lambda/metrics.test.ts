import { LambdaMetricFactory } from '../../../../src/monitoring/aws/lambda/metrics';

const DummyFunctionName = 'DummyFunctionName';

test('snapshot test: metricInvocations', () => {
  // GIVEN
  const unitToTest = new LambdaMetricFactory();

  // WHEN
  const metric = unitToTest.metricInvocations(DummyFunctionName);

  // THEN
  expect(metric.metricName).toStrictEqual('Invocations');
  expect(metric.dimensions).toStrictEqual({ FunctionName: DummyFunctionName });
  expect(metric).toMatchSnapshot();
});

test('snapshot test: metricDuration', () => {
  // GIVEN
  const unitToTest = new LambdaMetricFactory();

  // WHEN
  const metric = unitToTest.metricDuration(DummyFunctionName);

  // THEN
  Object.values(metric).forEach(eachMetric => {
    expect(eachMetric.metricName).toStrictEqual('Duration');
    expect(eachMetric.dimensions).toStrictEqual({ FunctionName: DummyFunctionName });
  });
  expect(metric).toMatchSnapshot();
});

test('snapshot test: metricErrors', () => {
  // GIVEN
  const unitToTest = new LambdaMetricFactory();

  // WHEN
  const metric = unitToTest.metricErrors(DummyFunctionName);

  // THEN
  expect(metric.metricName).toStrictEqual('Errors');
  expect(metric.dimensions).toStrictEqual({ FunctionName: DummyFunctionName });
  expect(metric).toMatchSnapshot();
});

test('snapshot test: metricThrottles', () => {
  // GIVEN
  const unitToTest = new LambdaMetricFactory();

  // WHEN
  const metric = unitToTest.metricThrottles(DummyFunctionName);

  // THEN
  expect(metric.metricName).toStrictEqual('Throttles');
  expect(metric.dimensions).toStrictEqual({ FunctionName: DummyFunctionName });
  expect(metric).toMatchSnapshot();
});
