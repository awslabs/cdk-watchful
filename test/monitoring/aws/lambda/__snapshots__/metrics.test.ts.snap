// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: metricDuration 1`] = `
Object {
  "avg": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "FunctionName": "DummyFunctionName",
    },
    "label": "Average",
    "metricName": "Duration",
    "namespace": "AWS/Lambda",
    "period": Duration {
      "amount": 5,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Average",
    "unit": undefined,
  },
  "max": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "FunctionName": "DummyFunctionName",
    },
    "label": "Maximum",
    "metricName": "Duration",
    "namespace": "AWS/Lambda",
    "period": Duration {
      "amount": 5,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Maximum",
    "unit": undefined,
  },
  "min": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "FunctionName": "DummyFunctionName",
    },
    "label": "Minimum",
    "metricName": "Duration",
    "namespace": "AWS/Lambda",
    "period": Duration {
      "amount": 5,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "Minimum",
    "unit": undefined,
  },
  "p50": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "FunctionName": "DummyFunctionName",
    },
    "label": "p50",
    "metricName": "Duration",
    "namespace": "AWS/Lambda",
    "period": Duration {
      "amount": 5,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "p50",
    "unit": undefined,
  },
  "p90": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "FunctionName": "DummyFunctionName",
    },
    "label": "p90",
    "metricName": "Duration",
    "namespace": "AWS/Lambda",
    "period": Duration {
      "amount": 5,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "p90",
    "unit": undefined,
  },
  "p99": Metric {
    "account": undefined,
    "color": undefined,
    "dimensions": Object {
      "FunctionName": "DummyFunctionName",
    },
    "label": "p99",
    "metricName": "Duration",
    "namespace": "AWS/Lambda",
    "period": Duration {
      "amount": 5,
      "unit": TimeUnit {
        "inMillis": 60000,
        "isoLabel": "M",
        "label": "minutes",
      },
    },
    "region": undefined,
    "statistic": "p99",
    "unit": undefined,
  },
}
`;

exports[`snapshot test: metricErrors 1`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": Object {
    "FunctionName": "DummyFunctionName",
  },
  "label": undefined,
  "metricName": "Errors",
  "namespace": "AWS/Lambda",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "Sum",
  "unit": undefined,
}
`;

exports[`snapshot test: metricInvocations 1`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": Object {
    "FunctionName": "DummyFunctionName",
  },
  "label": undefined,
  "metricName": "Invocations",
  "namespace": "AWS/Lambda",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "Sum",
  "unit": undefined,
}
`;

exports[`snapshot test: metricThrottles 1`] = `
Metric {
  "account": undefined,
  "color": undefined,
  "dimensions": Object {
    "FunctionName": "DummyFunctionName",
  },
  "label": undefined,
  "metricName": "Throttles",
  "namespace": "AWS/Lambda",
  "period": Duration {
    "amount": 5,
    "unit": TimeUnit {
      "inMillis": 60000,
      "isoLabel": "M",
      "label": "minutes",
    },
  },
  "region": undefined,
  "statistic": "Sum",
  "unit": undefined,
}
`;
